### DISCUSSION TITLE
js

### DISCUSSION THREAD

**Discussion by 1337kingOf** [gildas-lormeau/SingleFile#1116] [Q&A]
I already added extension, but how can it work
I need to execute ur extension like
self.driver.execute_script(run.SingleFile)

**Comment by 1337kingOf** [comment]
I read some info, or another way
How can i window.postMessage(data, "*"); to ur script?

**Comment by gildas-lormeau** [comment]
You have to port this script from JS to Python: https://github.com/gildas-lormeau/single-file-cli/blob/master/back-ends/webdriver-chromium.js.

**Comment by 1337kingOf** [comment]
> https://github.com/gildas-lormeau/single-file-cli/blob/master/back-ends/webdriver-chromium.js

How can i do that ?
Cause idk why, shortcuts dont work for save page
And only way execute ur script via js

**Comment by 1337kingOf** [comment]
can u give me example plz what i have to execute

**Comment by gildas-lormeau** [comment]
What is unclear to you in the code I suggested? It should be quite easy to port it in Python. Unfortunately, I don't use this programming language. I'm moving this issue in the "Discussions" tab because it's not describing a bug or a feature request.

**Reply by 1337kingOf** [reply]
where can i get 'options'?

**Reply by gildas-lormeau** [reply]
it's an object passed as parameter, it's built automatically from the flags passed from the CLI. You can consider it's valued by default to:
```js
{
  removeHiddenElements: true,
  removeUnusedStyles: true,
  removeUnusedFonts: true,
  removeImports: true,
  blockScripts: true,
  blockAudios: true,
  blockVideos: true,
  compressHTML: true,
  removeAlternativeFonts: true,
  removeAlternativeMedias: true,
  removeAlternativeImages: true,
  groupDuplicateImages: true
}
```

**Reply by 1337kingOf** [reply]
okay one thing what i cant understand 
how can i 
const scripts = await require("./common/scripts.js").get(options);
build this const with python 
like -subprocess.run(["node", "script.js"]?
or how?
Cause how i understand const scripts executing nodejs

**Reply by 1337kingOf** [reply]
or should i use 
import importlib.util

def get_scripts(options):
    spec = importlib.util.spec_from_file_location("scripts", "./common/scripts.js")
    scripts = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(scripts)
    return scripts.get(options)

async def main():
    options = ...  # define the options
    scripts = await get_scripts(options)
    # use the scripts

if __name__ == "__main__":
    asyncio.run(main())

**Reply by gildas-lormeau** [reply]
You have to port the code below

https://github.com/gildas-lormeau/single-file-cli/blob/a68e5297c5342078c5add0536ab4819d1346562d/back-ends/common/scripts.js#L66-L76

This function just concatenates SingleFile scripts actually.

The scripts to concatenate are:

https://github.com/gildas-lormeau/single-file-cli/blob/a68e5297c5342078c5add0536ab4819d1346562d/back-ends/common/scripts.js#L28-L32

**Reply by 1337kingOf** [reply]
```
from selenium import webdriver
import subprocess
import json

options = '''{
  removeHiddenElements: true,
  removeUnusedStyles: true,
  removeUnusedFonts: true,
  removeImports: true,
  blockScripts: true,
  blockAudios: true,
  blockVideos: true,
  compressHTML: true,
  removeAlternativeFonts: true,
  removeAlternativeMedias: true,
  removeAlternativeImages: true,
  groupDuplicateImages: true
}'''


def get_scripts(options):
    # Write the JavaScript code to a file
    with open("temp.js", "w") as f:
        f.write("""
const getScripts = async options => {
  return await require("./common/scripts.js").get(options);
};

getScripts(%s)
  .then(result => {
    console.log(JSON.stringify(result));
  })
  .catch(error => {
    console.error(error);
  });
""" % json.dumps(options))

    # Run the file using Node.js
    result = subprocess.run(["node", "temp.js"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout = result.stdout.decode("utf-8").strip()
    stderr = result.stderr.decode("utf-8").strip()
    print(stdout)
    print('------')
    print(stderr)

    # Check if there was an error
    if stderr:
        raise Exception(stderr)

    # Return the result
    return json.loads(stdout)


driver = webdriver.Chrome()
driver.get('https://yahoo.com')
driver.execute_script(get_scripts(options))
input(2222)
driver.quit()

```

**Reply by 1337kingOf** [reply]
or i have to do like 
```
driver.get('https://yahoo.com')
sss = '''
exports.get = async options => {
	let scripts = "let _singleFileDefine; if (typeof define !== 'undefined') { _singleFileDefine = define; define = null }";
	scripts += await readScriptFiles(SCRIPTS, basePath);
	scripts += await readScriptFiles(options && options.browserScripts ? options.browserScripts : [], "");
	if (options.browserStylesheets && options.browserStylesheets.length) {
		scripts += "addEventListener(\"load\",()=>{const styleElement=document.createElement(\"style\");styleElement.textContent=" + JSON.stringify(await readScriptFiles(options.browserStylesheets, "")) + ";document.body.appendChild(styleElement);});";
	}
	scripts += "if (_singleFileDefine) { define = _singleFileDefine; _singleFileDefine = null }";
	scripts += "(" + initSingleFile.toString() + ")();";
	return scripts;
};'''
driver.execute_script(get_scripts(options))

```

**Reply by 1337kingOf** [reply]
or u saying i have to read 
this 3 files
```
const SCRIPTS = [
	"lib/single-file.js",
	"lib/single-file-bootstrap.js",
	"lib/single-file-hooks-frames.js"
];
```
and execute them?

**Reply by gildas-lormeau** [reply]
You have to read, concatenate and execute them in the page as a script. See below:

https://github.com/gildas-lormeau/single-file-cli/blob/a68e5297c5342078c5add0536ab4819d1346562d/back-ends/webdriver-chromium.js#L117

**Reply by gildas-lormeau** [reply]
The function below executes SingleFile and returns the page data:

https://github.com/gildas-lormeau/single-file-cli/blob/a68e5297c5342078c5add0536ab4819d1346562d/back-ends/webdriver-chromium.js#L153-L168

**Reply by gildas-lormeau** [reply]
BTW, please note that you will have to publish your code with an AGPL license.

**Reply by 1337kingOf** [reply]
```
options = '''{
  removeHiddenElements: true,
  removeUnusedStyles: true,
  removeUnusedFonts: true,
  removeImports: true,
  blockScripts: true,
  blockAudios: true,
  blockVideos: true,
  compressHTML: true,
  removeAlternativeFonts: true,
  removeAlternativeMedias: true,
  removeAlternativeImages: true,
  groupDuplicateImages: true
}'''
a = ''
with open('../lib/single-file.js', 'r') as f:
    a += f.read()
    with open('../lib/single-file-bootstrap.js', 'r') as fs:
        a += fs.read()
        with open('../lib/single-file-hooks-frames.js', 'r') as fa:
            a += fa.read()
            driver.execute_script(a)

input(2222)


def get_page_data(driver, options):
    script = """
        const [options, callback] = arguments;
        getPageData()
            .then(pageData => callback({ pageData }))
            .catch(error => callback({ error: error && error.toString() }));
        async function getPageData() {
            const pageData = await singlefile.getPageData(options);
            if (options.includeInfobar) {
                await infobar.includeScript(pageData);
            }
            return pageData;
        }
    """
    result = driver.execute_async_script(script, json.loads(options))
    if "error" in result:
        raise Exception(result["error"])
    else:
        return result["pageData"]


print(get_page_data(driver, options))

input(3333)
```
correct?

**Reply by gildas-lormeau** [reply]
Yes, it seems correct.

**Reply by 1337kingOf** [reply]
no not working, i got bet way but 1 error

**Reply by gildas-lormeau** [reply]
What is the error?

**Comment by 1337kingOf** [comment]
```python
options = '''{
  removeHiddenElements: true,
  removeUnusedStyles: true,
  removeUnusedFonts: true,
  removeImports: true,
  blockScripts: true,
  blockAudios: true,
  blockVideos: true,
  compressHTML: true,
  removeAlternativeFonts: true,
  removeAlternativeMedias: true,
  removeAlternativeImages: true,
  groupDuplicateImages: true
}'''




def get_scripts(options):
    # Write the JavaScript code to a file
    with open("temp.js", "w") as f:
        f.write("""
const getScripts = async options => {
  return await require("./common/scripts.js").get(options);
};

getScripts(%s)
  .then(result => {
    console.log(JSON.stringify(result));
  })
  .catch(error => {
    console.error(error);
  });
""" % json.dumps(options))
    # Run the file using Node.js
    result = subprocess.run(["node", "temp.js"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout = result.stdout.decode("utf-8").strip()
    stderr = result.stderr.decode("utf-8").strip()
    print(stdout)
    print('------')
    print(stderr)

    # Check if there was an error
    if stderr:
        raise Exception(stderr)

    # Return the result
    return json.loads(stdout)
def get_page_data(driver, options):
    script = """
        const [options, callback] = arguments;
        getPageData()
            .then(pageData => callback({ pageData }))
            .catch(error => callback({ error: error && error.toString() }));
        async function getPageData() {
            const pageData = await singlefile.getPageData(options);
            if (options.includeInfobar) {
                await infobar.includeScript(pageData);
            }
            return pageData;
        }
    """
    result = driver.execute_async_script(script, options)
    if "error" in result:
        raise Exception(result["error"])
    else:
        return result["pageData"]


driver.execute_script(get_scripts(options))
input(2222)
print(get_page_data(driver, options))

input(3333)
```
and got error 
```
    raise Exception(result["error"])
Exception: TypeError: Cannot create property 'frames' on string '{
  removeHiddenElements: true,
  removeUnusedStyles: true,
  removeUnusedFonts: true,
  removeImports: true,
  blockScripts: true,
  blockAudios: true,
  blockVideos: true,
  compressHTML: true,
  removeAlternativeFonts: true,
  removeAlternativeMedias: true,
  removeAlternativeImages: true,
  groupDuplicateImages: true
}'
```

**Reply by gildas-lormeau** [reply]
`driver.execute_async_script(script, options)` seems to be incorrect, Could you do the same test with `driver.execute_async_script(script)` instead?

**Reply by 1337kingOf** [reply]
did it, and got 
 raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.TimeoutException: Message: script timeout

**Reply by gildas-lormeau** [reply]
Even on https://www.example.com?

**Reply by 1337kingOf** [reply]
yeah, i just tried

**Comment by 1337kingOf** [comment]
that i have in stdout 
[wow.txt](https://github.com/gildas-lormeau/SingleFile/files/10674115/wow.txt)

**Reply by gildas-lormeau** [reply]
This is the concatenated script...

**Reply by 1337kingOf** [reply]
does it correct ?
cause i 
driver.execute_script(get_scripts(options))
and he return return json.loads(stdout)

**Reply by 1337kingOf** [reply]
maybe i set incorrect  options ?

**Reply by gildas-lormeau** [reply]
Please do things step by step. First, just ensure you're able to inject the concatenated script in the page and verify ```singlefile.getPageData``` exists. It looks like this part in your code is not correct. I don't know the cause of the problem because I'm not a Python developer.

**Reply by 1337kingOf** [reply]
<img width="493" alt="Screen Shot 2023-02-07 at 12 54 51 PM" src="https://user-images.githubusercontent.com/110034904/217225755-9692e811-44ce-4d2e-be09-22365e1f374e.png">
<img width="821" alt="Screen Shot 2023-02-07 at 12 55 13 PM" src="https://user-images.githubusercontent.com/110034904/217225819-b7d12f9e-5484-4382-83fc-a72ef8756bc4.png">
it has

**Reply by 1337kingOf** [reply]
hey i need ur help asap bro plz

**Reply by gildas-lormeau** [reply]
It's not a SingleFile issue, it's a Selenium/Python issue. You showed that you were able to inject the script and that `singlefile.getPageData` exists. All you have to do is to call this function. You can maybe use the `debugger` keyword in JS to help you debugging this issue and understanding why you can't call it.

**Reply by gildas-lormeau** [reply]
Please also note that I do my best, I am not paid for this work, and I also have time to earn a living. So, please don't use words like "asap".

**Reply by 1337kingOf** [reply]
so u saying, if i run script singlefile.getPageData it have to save page?

**Reply by gildas-lormeau** [reply]
Yes, it will process the page and return you asynchronously an object containing the resulting HTML content.

**Reply by 1337kingOf** [reply]
<img width="550" alt="Screen Shot 2023-02-07 at 4 13 51 PM" src="https://user-images.githubusercontent.com/110034904/217268800-6b42b994-f983-404d-9380-606df84687c6.png">

**Reply by gildas-lormeau** [reply]
It works, as you can see `content` contains the expected HTML content.

**Comment by 1337kingOf** [comment]
<img width="269" alt="Screen Shot 2023-02-07 at 4 25 26 PM" src="https://user-images.githubusercontent.com/110034904/217271631-51d71963-d6bd-4e20-8f33-3625c59e5dcf.png">
<img width="546" alt="Screen Shot 2023-02-07 at 4 25 44 PM" src="https://user-images.githubusercontent.com/110034904/217271721-cfec74bd-64c3-4095-9581-1894b6192899.png">

i think content its my html file ?
but i got 
[123.txt](https://github.com/gildas-lormeau/SingleFile/files/10675997/123.txt)
rename from txt to html and u can see 
I need no save https://stackoverflow.com/questions/22761819/how-to-open-chrome-extension-page-programmatically

**Reply by 1337kingOf** [reply]
this script not saving as if i save like 
<img width="683" alt="Screen Shot 2023-02-07 at 4 31 47 PM" src="https://user-images.githubusercontent.com/110034904/217273333-d1979c6d-198a-49b6-bfe3-abd5f9108ace.png">

**Reply by 1337kingOf** [reply]
and does it norm when there stats:underfined ?

**Reply by gildas-lormeau** [reply]
Some resources are blocked by CORS. Try to launch Chrome with the flag `--disable-web-security`. It's normal that `stats` is empty because you did not enable them.

**Reply by 1337kingOf** [reply]
but remember im now using cli, im using as python selenium

**Reply by 1337kingOf** [reply]
how can i run script as extension default ?

**Reply by gildas-lormeau** [reply]
See https://github.com/gildas-lormeau/single-file-cli/blob/dd07c0b15b3305b87665c154d9e895fff0542c80/back-ends/webdriver-chromium.js#L64

**Reply by 1337kingOf** [reply]
wow bro u god, its works fineee, so how can i disable js, or replace urls?

**Reply by 1337kingOf** [reply]
hey i got error 
selenium.common.exceptions.JavascriptException: Message: javascript error: Failed to set the 'textContent' property on 'Node': This document requires 'TrustedScript' assignment.

**Reply by 1337kingOf** [reply]
this document requires ‘trustedscript’ assignment.
how can i fix it ?

**Reply by gildas-lormeau** [reply]
I don't know

**Reply by 1337kingOf** [reply]
this error comes when i execute await singlefile.getPageData() on gmail site

**Reply by 1337kingOf** [reply]
instead of, when i save gmail with extension, its save successfully 
But when i save with selenium its trhrow error, i think we need set options

**Reply by gildas-lormeau** [reply]
Use something else than selenium, like pyppeteer or playwright.

**Reply by 1337kingOf** [reply]
bro but error not in selenium

**Reply by 1337kingOf** [reply]
i tried to do manully, load script and run get page , its throw same error, instead of i use extension and save page

**Reply by 1337kingOf** [reply]
and favions dont loading bro, idk why, on extension its loading , on cli - not

**Reply by gildas-lormeau** [reply]
Because the favico URL returns a 404 maybe? Maybe, that's normal....

**Reply by 1337kingOf** [reply]
but im saying to u, when i save with extension i dont have problems, when i save with cli, i have those problems, i cant see code and thats why i ask u, u had obfuscate code bro

**Reply by 1337kingOf** [reply]
i need help plz

**Reply by gildas-lormeau** [reply]
I won't do your work for free, I'm sorry.

**Reply by 1337kingOf** [reply]
should i pay?
